Support Outlook 2016 signature editor

michaelDCurran opened the pull request:
Fixes #7253

michaelDCurran committed:
MS Office 2016 now has a native UI Automation implementation for _WwN controls (Outlook signature editor, spell check field etc), so use this if available. However, move previous _WwN support specifically to IAccessible, as when it is at the Window level it would conflict with UI Automation.
This change is most important for supporting the Outlook 2016 Signature editor.

feerrenrut commented on source/speech.py:
```diff
@@ -313,7 +313,11 @@ def speakObjectProperties(obj,reason=controlTypes.REASON_QUERY,index=None,**allo
+	isEditable=(reason!=controlTypes.REASON_FOCUSENTERED and obj.TextInfo!=NVDAObjectTextInfo and (role in (controlTypes.ROLE_EDITABLETEXT,controlTypes.ROLE_TERMINAL,controlTypes.ROLE_DOCUMENT) or controlTypes.STATE_EDITABLE in obj.states))
+	# Browse mode document treeInterceptors currently handle announcing text content themselves, thus we should not class these as editable for deciding whether to announce text content here.
+	import browseMode
+	if isEditable and isinstance(obj.treeInterceptor,browseMode.BrowseModeDocumentTreeInterceptor) and obj is obj.treeInterceptor.rootNVDAObject:
```
The logic for `isEditable` is pretty hard to digest. I think it would aid readability if it was broken up a bit, perhaps moved to an self contained function. Comments on what each of the escapes from being editable would also help, perhaps some examples of what these things mean.

michaelDCurran committed:
Abstract rules for when navigable text should be presented in speech and braille into a _isNavigableText property on NVDAObjects. This property is currently private as the name could change.
The rules were taken from Braille, but the checking of no readonly state for documents has been removed as this was only to protect against browse mode, which is already handled higher up in Braille.

feerrenrut commented on source/speech.py:
```diff
@@ -313,11 +313,9 @@ def speakObjectProperties(obj,reason=controlTypes.REASON_QUERY,index=None,**allo
 	import browseMode
-	if isEditable and isinstance(obj.treeInterceptor,browseMode.BrowseModeDocumentTreeInterceptor) and obj is obj.treeInterceptor.rootNVDAObject:
-		isEditable=False
+	isEditable=(not isinstance(obj.treeInterceptor,browseMode.BrowseModeDocumentTreeInterceptor) or obj.treeInterceptor.passThrough) and obj._hasNavigableText
```
I would be tempted to extract the first part of the `and` to a `isBrowseMode()` method

feerrenrut reviewed (approved):
Overall looks good to me.

michaelDCurran committed:
Fix regression introduced by #7253 that was causing double speech as NVDAObjectTextInfos were being classed as navigable. NVDAObjecTextInfo is now never classed as navigable text, and speech also correctly again checks to make sure the reason is not focusEntered.

jcsteh commented:
This seems to be causing a regression when pressing NVDA+tab on an editable text field in focus mode. isEditable returns False in this case due to the DocumentTreeInterceptor check, so the value gets read instead of the text, placeholder isn't read, etc.

STR for first case:

1. Open this URL in Chrome or Firefox: `data:text/html,<input type="text" aria-label="foo" aria-placeholder="bar">`
2. Move to the edit field in browse mode.
3. Press NVDA+tab.
    * Expected: "foo  edit focused, bar, blank"
    * Actual: "foo  edit focused"

STR for second case:

1. Open this URL in Firefox: `data:text/html,<div contentEditable="true">foo</div>`
2. Move to the edit field in browse mode.
3. Press NVDA+tab.
    * Expected: "section editable focused, foo"
    * Actual: "section editable focused"

michaelDCurran committed:
Merge branch 'master' into i7253

michaelDCurran committed:
speech.speakObject: ensure we only drop reporting of text content  for objects in treeInterceptors if they are actually the root of the treeInterceptor. Also break up the logic into commented lines.
